app/controllers/concerns/course/lesson_plan/strategies/fomo_personalization_strategy.rb

```diff
@@ -0,0 +1,127 @@
+# frozen_string_literal: true
+class FomoPersonalizationStrategy < BasePersonalizationStrategy
+  MIN_LEARNING_RATE = 0.67
+  MAX_LEARNING_RATE = 1.0
+  HARD_MIN_LEARNING_RATE = 0.5
+  DATE_ROUNDING_THRESHOLD = 0.8
+
+  # Shifts start_at of relevant lesson plan items and resets the bonus_end_at and end_at
+  # of the same items. The amount shifted is based the learning rate precomputed.
+  #
+  # The expected precomputed_data is the default data from precompute_data.
+  #
+  # @param [CourseUser] course_user The user to adjust the personalized timeline for.
+  # @param [Hash] precomputed_data The default data precomputed by precompute_data.
+  def execute(course_user, precomputed_data) # rubocop:disable Metrics/AbcSize, Metrics/MethodLength
+    return if precomputed_data[:learning_rate_ema].nil?
+
+    @course_tz = course_user.course.time_zone
+    reference_point = personal_point = precomputed_data[:items].first.reference_time_for(course_user).start_at
+    course_user.transaction do
+      precomputed_data[:items].each do |item|
+        reference_point, personal_point = update_points(course_user, item, precomputed_data[:submitted_items],
+                                                        reference_point, personal_point)
+        next if should_skip_item(course_user, item, precomputed_data[:submitted_items])
+
+        reference_time = item.reference_time_for(course_user)
+        personal_time = item.find_or_create_personal_time_for(course_user)
```

Comment:
This is twice (another one in line 78 below). Won't this call an extra query here?

Reply:
Yep, I've thought about it in terms of the tradeoff between code quality/readability/abstraction vs a constant factor i.e. 2n queries instead of n. I previously leaned towards the former as I felt that n wouldn't really be very significant, given that it's an n that goes through a lot of filtering. But now I'm thinking it might be better to keep the queries at n. What do you think?

Reply:
I think lets try to keep it at n queries where possible.. haha well it may not be significant now but next time it may pile up when we keep adding more and more